cpp/oneapi/dal/backend/primitives/distance/squared_l2_distance_auxilary_dpc.cpp

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/backend/primitives/distance/squared_l2_distance_auxilary.hpp"
+
+#include "oneapi/dal/backend/primitives/blas.hpp"
+#include "oneapi/dal/backend/primitives/reduction.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+template <typename Float>
+sycl::event compute_squared_l2_norms(sycl::queue& q,
+                                     const ndview<Float, 2>& inp,
+                                     ndview<Float, 1>& out,
+                                     const event_vector& deps) {
+    return reduce_by_rows(q, inp, out, sum<Float>{}, square<Float>{}, deps);
+}
+
+template <typename Float>
+std::tuple<array<Float>, sycl::event> compute_squared_l2_norms(sycl::queue& q,
+                                                               const ndview<Float, 2>& inp,
+                                                               const event_vector& deps,
+                                                               const sycl::usm::alloc& alloc) {
+    const auto n_samples = inp.get_dimension(0);
+    auto res_array = array<Float>::empty(q, n_samples, alloc);
```

Comment:
Why not ndarray?

Reply:
I suppose it's better to pass pimpl entity instead.

Reply:
Fixed